lib/src/renamer.dart

```diff
@@ -0,0 +1,259 @@
+  /// as input.
+  ///
+  /// [keys] is the list of keys that can be referenced in [code] and must
+  /// all appear in every input map.
+  static Renamer<Map<String, String>> map(String code, List<String> keys) {
+    return Renamer(code, {for (var key in keys) key: (input) => input[key]});
+  }
+
+  /// Creates a Renamer based on [code] and a map from keys to functions that
+  /// take an input and return a string.
+  ///
+  /// All keys must consist of only lowercase letters, underscores, and hyphens.
+  ///
+  /// If provided, [sourceUrl] will appear in parsing errors. It can be
+  /// a [String] or a [Uri].
+  factory Renamer(String code, Map<String, String Function(T input)> keys,
+      {dynamic sourceUrl}) {
+    for (var key in keys.keys) {
+      if (!RegExp(r'^[a-z_-]*$').hasMatch(key)) {
+        throw ArgumentError(
+            'Invalid key "$key". Must use only lowercase letters, '
+            'underscores, and hyphens.');
+      }
+    }
+    var scanner = StringScanner(code, sourceUrl: sourceUrl);
+    var statements = <_Statement<T>>[];
+    scanner.scan(_statementDelimiter);
+    while (!scanner.isDone) {
+      statements.add(_readStatement(scanner, keys));
+    }
+    return Renamer._(keys, statements);
+  }
+
+  /// Reads the next statement (and the trailing delimiter, if any) from
+  /// [scanner].
+  static _Statement<T> _readStatement<T>(
+      StringScanner scanner, Map<String, String Function(T input)> keys) {
+    var start = scanner.position;
+    FormatException lastException;
+    // Tries each key in succession until one is successfully returned.
+    for (var entry in keys.entries) {
+      try {
+        scanner.position = start;
+        var statement = _tryKey(scanner, entry.key, entry.value);
+        if (statement != null) return statement;
+      } on FormatException catch (e) {
+        lastException = e;
+      }
+    }
+    if (lastException == null) {
+      scanner.error('invalid key');
+    }
+    throw lastException;
+  }
+
+  /// Tries to read a statement for [key] from [scanner].
+  ///
+  /// If [key] is non-null and the next text in [scanner] is not that key, this
+  /// returns null immediately, since the parse error would not be useful.
+  ///
+  /// Otherwise, after consuming the key clause (if any), attempts to read
+  /// the matcher and output clauses, throwing if it's unable to.
+  static _Statement<T> _tryKey<T>(
+      StringScanner scanner, String key, String Function(T input) keyFunction) {
+    if (key.isNotEmpty && !scanner.scan('$key ')) return null;
+    var matcher = _readMatcher(scanner);
+    scanner.expect(' to ');
+    var output = _readOutput(scanner);
+    return _Statement(keyFunction, matcher, output);
+  }
+
+  /// Reads a matcher clause and its trailing space from [scanner].
+  static RegExp _readMatcher(StringScanner scanner) {
+    int quote;
+    if ({$single_quote, $double_quote}.contains(scanner.peekChar())) {
+      quote = scanner.readChar();
+    }
```

Comment:
I think people who are used to JS or shell idioms would get pretty confused using quotes to delimit a regular expression. I'd just stick with allowing spaces to be backslash-escaped, especially since in the use-cases we know of we expect it to be very rare that someone would want to write a space in the input or output. As a bonus, that'll also make this logic simpler :smile:.

Reply:
Removed